File: src/tools/steamtools.py
from typing import List
from datetime import datetime

from .base import UnlockTool
from ..models import DepotInfo, SteamAppInfo


class SteamTools(UnlockTool):
    """SteamTools解锁工具实现"""

    async def setup(
        self,
        depot_data: List[DepotInfo],
        app_info: SteamAppInfo,
    ) -> bool:
        """设置SteamTools解锁"""
        st_path = self.steam_path / "config" / "stplug-in"
        st_path.mkdir(parents=True, exist_ok=True)

        lua_content = f"""
-- Generated Lua Manifest by Onekey
-- Steam App {app_info.appId} Manifest
-- Name: {app_info.name}
-- Generated: {datetime.now().strftime("%Y-%m-%d %H:%M:%S")}
-- Total Depots: {app_info.depotCount}
-- Total DLCs: {app_info.dlcCount}

-- MAIN APP
addappid({app_info.appId}, "0", "{app_info.workshopDecryptionKey}")

-- ALL Depots
"""

        for depot in depot_data:
            lua_content += (
                f'addappid({depot.depot_id}, "1", "{depot.decryption_key}")\n'
            )

        lua_file = st_path / f"{app_info.appId}.lua"
        lua_file.write_text(lua_content, encoding="utf-8")

        return True

File: internal/steamtools/steamtools.go
package steamtools

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"onekey/internal/models"
)

// Setup generates the SteamTools Lua unlock file in Steam's stplug-in directory.
func Setup(steamPath string, appInfo *models.SteamAppInfo, manifests []models.ManifestInfo) error {
	stPath := filepath.Join(steamPath, "config", "stplug-in")
	if err := os.MkdirAll(stPath, 0755); err != nil {
		return fmt.Errorf("create stplug-in directory: %w", err)
	}

	var b strings.Builder

	fmt.Fprintf(&b, "-- Generated Lua Manifest by Onekey\n")
	fmt.Fprintf(&b, "-- Steam App %s Manifest\n", appInfo.AppID)
	fmt.Fprintf(&b, "-- Name: %s\n", appInfo.Name)
	fmt.Fprintf(&b, "-- Generated: %s\n", time.Now().Format("2006-01-02 15:04:05"))
	fmt.Fprintf(&b, "-- Total Depots: %d\n", appInfo.DepotCount)
	fmt.Fprintf(&b, "-- Total DLCs: %d\n", appInfo.DLCCount)
	fmt.Fprintf(&b, "\n-- MAIN APP\n")
	fmt.Fprintf(&b, "addappid(%s, \"0\", \"%s\")\n", appInfo.AppID, appInfo.WorkshopDecryptionKey)
	if appInfo.AccessToken != "" && appInfo.AccessToken != "0" {
		fmt.Fprintf(&b, "addtoken(%s, \"%s\")\n", appInfo.AppID, appInfo.AccessToken)
	}
	fmt.Fprintf(&b, "\n-- ALL Depots\n")

	// Deduplicate depots: each depot_id should appear only once.
	seen := make(map[string]bool)
	for _, m := range manifests {
		if seen[m.DepotID] {
			continue
		}
		seen[m.DepotID] = true
		fmt.Fprintf(&b, "addappid(%s, \"1\", \"%s\")\n", m.DepotID, m.DepotKey)
	}

	luaFile := filepath.Join(stPath, appInfo.AppID+".lua")
	if err := os.WriteFile(luaFile, []byte(b.String()), 0644); err != nil {
		return fmt.Errorf("write lua file: %w", err)
	}

	return nil
}
